File: algorigthms/programming-pearls/01-bitMapSorting.py
'''
有 10^7 个惟一数字存在一个文件中，规定只能使用不超过 1MB 的内存来完成排序怎么实现？
用  bitMap 来完成排序。
Precise Problem Statement.
Input 	A file containing at most n positive integers, each less than n, where n = 10^7.
				It is a fatal error if any integer occurs twice in the input. No other data is associated
				with the integer.

Output 	A sorted list in increasing order of the input integers.

Constraints 	At most(roughly) a megabyte of storage is available in main memory; ample disk storage is 
							available. The run time can be at most serveral minutes; a run time of ten seconds need not 
							be decreased.
'''
from bitarray import bitarray
import random

''' 下面用一个简单数据集来完成测试。'''

MAX_SIZE = 1024

def generate_data (len):
	''' 随机生成输入数据'''
	return random.sample(range(len), len)

def bit_map_sort(data):
	bit_map = bitarray(MAX_SIZE)
	bit_map.setall(0)
	for d in data:
		bit_map[d] = 1
	
	data = []
	i = -1
	for b in bit_map:
		i += 1
		if b: data.append(i)

	return data

bm = bitarray(10)
bm.setall(0)
print(bm[0], bm[1], bm)
data = generate_data(10)
print('Before sorting: ')
print(data)
print('After sorting: ')
data = bit_map_sort(data)
print(data)

data = [23, 68, 90, 223, 12, 33, 22, 110]
print('\nBefore sorting: ')
print(data)
data = bit_map_sort(data)
print('After sorting:')
print(data)

File: algorigthms/programming-pearls/01-bitMapSorting.cpp
#include <iostream>
#include <bitset>

#define MAX_SIZE  1024

using namespace std;

void bitMapSort(int * data, int size)
{
	bitset<MAX_SIZE> bitMap;
	for (int i = 0; i < size; ++i) {
		bitMap.set(data[i], 1);
	}
	int j = 0;
	for (int i = 0; i < MAX_SIZE; ++i) {
		if (bitMap[i]) {
			data[j++] = i;
		}
	}
}

void printData(int * data, int size)
{
	for (int i = 0; i < size; ++i) {
		cout << data[i] << "\t";
	}
}

int main ()
{
	int data[] = {79, 75, 85, 7, 45, 20, 81, 72, 6, 33, 53, 4, 19, 63, 26, 39, 27, 43, 22, 11, 24, 61, 60, 44, 25, 86, 66, 3, 92, 48};
	int size = sizeof(data) / sizeof(data[0]);
	cout << "Before sorting: ";
	printData(data, size);
	bitMapSort(data, size);
	cout << "\n After sorting: ";
	printData(data, size);
	return 0;
}